models/issue_indexer.go

```diff
@@ -0,0 +1,201 @@
+		numericQuery(repoID, "RepoID"),
+		bleve.NewDisjunctionQuery(
+			bleve.NewPhraseQuery(fields, "Title"),
+			bleve.NewPhraseQuery(fields, "Content"),
+		))
+	search := bleve.NewSearchRequestOptions(indexerQuery, 2147483647, 0, false)
+	search.Fields = []string{"ID"}
+
+	index, err := bleve.Open(setting.IssueIndexerPath)
+	if err != nil {
+		return nil, err
+	}
+	result, err := index.Search(search)
+	if err != nil {
+		index.Close()
+		return nil, err
+	}
+	if err = index.Close(); err != nil {
+		return nil, err
+	}
+
+	issueIDs := make([]int64, len(result.Hits))
+	for i, hit := range result.Hits {
+		issueIDs[i] = int64(hit.Fields["ID"].(float64))
+	}
+	return issueIDs, nil
+}
+
+// InitIssuesIndexer initialize issue indexer
+func InitIssuesIndexer() error {
+	_, err := os.Stat(setting.IssueIndexerPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = CreateIssuesIndexer(); err != nil {
+				return err
+			}
+			go func() {
+				if err := PopulateIssuesIndexer(); err != nil {
+					log.Error(4, "PopulateIssuesIndex: %v", err)
+				}
+			}()
+		} else {
+			return err
+		}
+	}
+	indexerUpdateQueue = make(chan *Issue, 10)
+	go processIndexerUpdateQueue()
+	return nil
+}
+
+// CreateIssuesIndexer create an issues indexer
+func CreateIssuesIndexer() error {
+	mapping := bleve.NewIndexMapping()
+	docMapping := bleve.NewDocumentMapping()
+
+	docMapping.AddFieldMappingsAt("ID", bleve.NewNumericFieldMapping())
+	docMapping.AddFieldMappingsAt("RepoID", bleve.NewNumericFieldMapping())
+
+	textFieldMapping := bleve.NewTextFieldMapping()
+	textFieldMapping.Analyzer = simple.Name
+	docMapping.AddFieldMappingsAt("Title", textFieldMapping)
+	docMapping.AddFieldMappingsAt("Content", textFieldMapping)
+
+	mapping.AddDocumentMapping("issues", docMapping)
+	index, err := bleve.New(setting.IssueIndexerPath, mapping)
+	if err != nil {
+		return err
+	}
+	return index.Close()
+}
+
+// PopulateIssuesIndexer write all issues in the database to the issues indexer
+func PopulateIssuesIndexer() error {
+	index, err := bleve.Open(setting.IssueIndexerPath)
+	if err != nil {
+		return err
+	}
+	for page := 1; ; page++ {
+		repos, err := Repositories(&SearchRepoOptions{
+			Page: page,
+			PageSize: 10,
+		})
+		if err != nil {
+			index.Close()
+			return err
+		}
+		if len(repos) == 0 {
+			return index.Close()
+		}
+		batch := index.NewBatch()
+		for _, repo := range repos {
+			issues, err := Issues(&IssuesOptions{
+				RepoID: repo.ID,
+				IsClosed: util.OptionalBoolNone,
+				IsPull: util.OptionalBoolNone,
+				Page: -1, // do not page
+			})
+			if err != nil {
+				index.Close()
+				return err
+			}
+			for _, issue := range issues {
+				err = batch.Index(issue.indexUID(), issue.issueData())
+				if err != nil {
+					index.Close()
+					return err
+				}
+			}
+		}
+		index.Batch(batch)
+	}
+}
+
+func processIndexerUpdateQueue() {
+	for {
+		select {
+		case issue := <-indexerUpdateQueue:
+			index, err := bleve.Open(setting.IssueIndexerPath)
```

Comment:
Only one thread can open an indexer at a time. If the update thread always held the indexer open, no other threads could ever access the indexer. I'll add a comment